static/src/javascripts/projects/common/modules/crosswords/crossword.js

```diff
@@ -0,0 +1,772 @@
+            });
+        }
+    }
+
+    check(entry: Object): void {
+        const cells = helpers.cellsForEntry(entry);
+
+        if (entry.solution) {
+            const badCells = zip(cells, entry.solution)
+                .filter(cellAndSolution => {
+                    const coords = cellAndSolution[0];
+                    const cell = this.state.grid[coords.x][coords.y];
+                    const solution = cellAndSolution[1];
+                    return (
+                        /^[A-Z]$/.test(cell.value) && cell.value !== solution
+                    );
+                })
+                .map(cellAndSolution => cellAndSolution[0]);
+
+            this.setState({
+                grid: helpers.mapGrid(this.state.grid, (cell, gridX, gridY) => {
+                    if (
+                        badCells.some(bad => bad.x === gridX && bad.y === gridY)
+                    ) {
+                        cell.isError = true;
+                        cell.value = '';
+                    }
+
+                    return cell;
+                }),
+            });
+
+            setTimeout(() => {
+                this.setState({
+                    grid: helpers.mapGrid(
+                        this.state.grid,
+                        (cell, gridX, gridY) => {
+                            if (
+                                badCells.some(
+                                    bad => bad.x === gridX && bad.y === gridY
+                                )
+                            ) {
+                                cell.isError = false;
+                                cell.value = '';
+                            }
+
+                            return cell;
+                        }
+                    ),
+                });
+            }, 150);
+        }
+    }
+
+    hiddenInputValue(): string {
+        const cell = this.state.cellInFocus;
+
+        let currentValue;
+
+        if (cell) {
+            currentValue = this.state.grid[cell.x][cell.y].value;
+        }
+
+        return currentValue || '';
+    }
+
+    hasSolutions(): boolean {
+        return 'solution' in this.props.data.entries[0];
+    }
+
+    isHighlighted(x: number, y: number): boolean {
+        const focused = this.clueInFocus();
+        return focused
+            ? focused.group.some(id => {
+                  const entry = this.props.data.entries.find(e => e.id === id);
+                  return helpers.entryHasCell(entry, x, y);
+              })
+            : false;
+    }
+
+    render(): Element {
+        const focused = this.clueInFocus();
+
+        const anagramHelper =
+            this.state.showAnagramHelper &&
+            React.createElement(AnagramHelper, {
+                crossword: this,
+                focussedEntry: focused,
+                entries: this.props.data.entries,
+                grid: this.state.grid,
+                close: this.onToggleAnagramHelper,
+            });
+
+        return React.createElement(
```

Comment:
Aah React, post jsx conversion

Reply:
Yea, this did make me wonder if there's a reason we're not using JSX for this.

Reply:
I guess we didn't have a transpile step back when crosswords was implemented. I think @MatthewJWalls is considering replacing React with Preact at some point. Maybe we can look at using JSX then

Reply:
That sounds grand, @SiAdcock

Reply:
we did have – we had remove it when we dropped JSPM and i had to manually convert all the jsx to this...a dreadful time

Reply:
You have my condolences.